### PR TITLE
WIP: relay dedupe/debounce

### PR DESCRIPTION
**Pull request by matthieu-foucault**
DO NOT MERGE

This is making the batch middleware very unhappy, but it works.
This needs to be changed to handle debouncing before the batch middleware instead of modifying the request it creates

### CODE REVIEW

**Comment by wenzowski** [comment]
This looks really cool and it's awesome that this works. In practice, this currently violates the `Disposable` interface exposed by [`commitMutation`](https://github.com/facebook/relay/blob/master/packages/relay-runtime/mutations/commitMutation.js#L80) since the [`Disposable type`](https://github.com/facebook/relay/blob/master/packages/relay-runtime/util/RelayRuntimeTypes.js#L24) doesn't allow registration of a continuation to inform the caller when the Disposable has been `dispose()`'d. As a result, we really want to split this into two parts: 
- a debounce/throttle-based "delay" middleware at the network layer that releases all queued mutations no more frequently than at a specific interval but never interferes with their being dispatched, and
- a singleton mixin for use in mutation modules that automatically preserves a reference to the last mutation and calls `dispose()` on it if it has not yet resolved via a `onCompleted` or `onError` continuation.
This way, each mutation is responsible for managing its own cancellation state and unnecessary mutations are given an opportunity to be cancelled by way of the delay introduced at the network level, without potentially interfering with any mutations that should not be `dispose()`'d of.

**Comment by matthieu-foucault** [comment]
@wenzowski Looking at your comment above, I think that one small adjustment that needs to be made is that the network layer middleware does need to interfere with the mutations being dispatched, since `dispose()` will not cancel the network request.
As per the only docs I could find on it (https://relay.dev/docs/en/api-reference#return-value-6)
> Calling disposable.dispose() will revert the optimistic update, and Relay won’t update the store or call any success/error callback, but the network request is not guaranteed to be cancelled.

File: app/cypress/integration/email.spec.js [hunk 1]
```diff
@@ -41,13 +41,12 @@ function requestAccessToOrg() {
   cy.visit('/reporter');
   cy.get('#page-content');
   cy.get('button#org-dropdown').click();
-  cy.wait(500);
   cy.get('.dropdown-menu.show a.dropdown-item')
     .contains('MacDonalds Agriculture, Ltd.')
     .click();
-  cy.wait(500);
+  cy.get('.card h4').contains('MacDonalds Agriculture, Ltd.');
```
Comment (wenzowski) [inline app/cypress/integration/email.spec.js:47]: would be awesome if in refactoring we got some semantic css classes in...

File: app/mutations/BaseMutation.ts [hunk 2]
```diff
@@ -53,10 +53,25 @@ export default class BaseMutation<T extends BaseMutationType = never> {
       }
     ) {
       return new Promise<T['response']>((resolve, reject) => {
-        commitMutationDefault<T>(environment, {
+        // Debounced mutations should be commited immediately to perform the optimisticUpdate
+        // The actual request will be cancelled in the network layer
```
Comment (wenzowski) [inline app/mutations/BaseMutation.ts:57]: it's only getting deferred in the network layer, right? Up to 250 ms?

File: app/lib/relay-environment/debounce-mutations.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+} from 'react-relay-network-modern/node8';
+import {CacheConfigWithDebounce} from 'next-env';
+
+let debouncedMutation: {
+  debounceKey: string;
+  debouncedFn: () => void;
+  timeoutId: number;
+} = null;
+
+const debounceMutationMiddleware = (timeout = 250): Middleware => {
```
Comment (wenzowski) [inline app/lib/relay-environment/debounce-mutations.ts:14]: love `timeout` as a good default

File: app/mutations/BaseMutation.ts [hunk 4]
```diff
@@ -53,10 +53,25 @@ export default class BaseMutation<T extends BaseMutationType = never> {
...
+        // Here we either dispose of a debounced mutation, or remove it from the map when it errors/completes
+        if (debounceKey) {
+          const previousMutation = debouncedMutationMap.get(debounceKey);
+          if (previousMutation) {
+            previousMutation.dispose();
```
Comment (wenzowski) [inline app/mutations/BaseMutation.ts:62]: perfect! We talked about some hiccups with disposables not getting fully disposed...did we resolve the edge case of disposing network-requests that are already in flight?

File: app/tests/unit/mutations/application/createApplicationMutation.test.ts [hunk 5]
```diff
@@ -12,7 +12,6 @@ const mutation = `
     $input: CreateApplicationMutationChainInput!
   ) {
     createApplicationMutationChain(input: $input) {
-      clientMutationId
```
Comment (wenzowski) [inline app/tests/unit/mutations/application/createApplicationMutation.test.ts:15]: Is clientMutationId no longer needed for relay's optimism? It's just an echoback field: it's a no-op server-side.

File: docs/stack.md [hunk 6]
```diff
@@ -100,6 +100,33 @@ Other tools such as alembic ([https://alembic.sqlalchemy.org/](https://alembic.s
 
 #### Debouncing Mutations
 
+If we expect a mutation to be dispatched several times in quick succession, and if the last instance of the mutation is the only one required to generate the correct state, then only the final mutation should be sent to the back-end, to reduce the load and network traffic.
+This is the case for instance for mutations that are sent on every key press in an input.
```
Comment (wenzowski) [inline docs/stack.md:104]: Only idempotent mutations are candidates for debounce.

Comment (wenzowski) [reply docs/stack.md:104]: Some mutations that need to be dispatched in quick succession are a better fit for the batch dispatcher, eg. a counter mutation.

File: app/tests/unit/mutations/application/createApplicationMutation.test.ts (hunk 5 above)
Comment (matthieu-foucault) [reply app/tests/unit/mutations/application/createApplicationMutation.test.ts:15]: It's not needed since https://github.com/facebook/relay/releases/tag/v8.0.0

File: docs/stack.md (hunk 6 above)
Comment (matthieu-foucault) [reply docs/stack.md:104]: Will update the docs for completeness. Regarding the batch dispatcher, should it be opt-in? All mutations are batched (with a 10ms window)

File: app/mutations/BaseMutation.ts (hunk 4 above)
Comment (matthieu-foucault) [reply app/mutations/BaseMutation.ts:62]: Right. If the mutation made it past the `debounceMiddleware` then we may dispose of a mutation that will make its way to the serve. I guess the milddeware should remove the mutation from the `debouncedMutationMap` before calling `next()`. Will update that
